internal/statemachine/common_states.go

```diff
@@ -133,66 +133,128 @@ func (stateMachine *StateMachine) generateDiskInfo() error {
+				stateMachine.RootfsSize, desiredSize,
+			)
+		} else {
 			stateMachine.RootfsSize = desiredSize
 		}
 	}
 
-	stateMachine.syncGadgetStructureRootfsSize()
+	if rootfsStructure != nil {
+		rootfsStructure.Size = stateMachine.RootfsSize
+	}
+
 	return nil
 }
 
-// findRootfsVolume finds the volume associated to the rootfs
-func (stateMachine *StateMachine) findRootfsVolume() (*gadget.Volume, string) {
+// getRootfsMinSize gets the minimum size needed to hold the built rootfs directory
+func (stateMachine *StateMachine) getRootfsMinSize() (quantity.Size, error) {
+	rootfsMinSize, err := helper.Du(stateMachine.tempDirs.rootfs)
+	if err != nil {
+		return quantity.Size(0), fmt.Errorf("Error getting rootfs size: %s", err.Error())
+	}
+
+	// Take into account ext4 filesystems metadata size
+	rootfsPadding := ext4Padding
+	rootfsMinSize = quantity.Size(math.Ceil(float64(rootfsMinSize) * ext4FudgeFactor))
+	rootfsMinSize += rootfsPadding
+
+	return stateMachine.alignToSectorSize(rootfsMinSize), nil
+}
+
+// getGadgetRootfsMinSize gets the minimum size of the rootfs requested in the
+// gadget YAML
+func (stateMachine *StateMachine) getGadgetRootfsMinSize(rootfsStructure *gadget.VolumeStructure) quantity.Size {
+	if rootfsStructure == nil {
+		return quantity.Size(0)
+	}
+
+	return stateMachine.alignToSectorSize(rootfsStructure.MinSize)
+}
+
+// getRootfsDesiredSize subtracts the size and offsets of the existing structures
+// from the requested image size to determine how much room is left for the rootfs
+func (stateMachine *StateMachine) getRootfsDesiredSize(rootfsVolume *gadget.Volume, rootfsVolumeName string) (quantity.Size, bool) {
+	if rootfsVolume == nil {
+		return quantity.Size(0), false
+	}
+
+	desiredSize, found := stateMachine.ImageSizes[rootfsVolumeName]
+	if !found {
+		// So far we do not know of a desired size for the rootfs
+		return quantity.Size(0), found
+	}
+
+	reservedSize := calculateNoRootfsSize(rootfsVolume)
+	desiredSize = helper.SafeQuantitySubtraction(desiredSize, reservedSize)
+	desiredSize = stateMachine.alignToSectorSize(desiredSize)
+
+	return desiredSize, found
+}
+
+// calculateNoRootfsSize determines the needed space for existing structures
+// except for the rootfs
+func calculateNoRootfsSize(v *gadget.Volume) quantity.Size {
+	var size quantity.Size
+	for _, s := range v.Structure {
+		if isRootfsStructure(&s) { //nolint:gosec,G301
+			continue
+		}
+		if s.Offset != nil && quantity.Size(*s.Offset) > size {
+			size = quantity.Size(*s.Offset)
+		}
+		size += s.MinSize
+	}
+
+	return size
+}
+
+// findRootfsVolumeStructure finds the volume and the structure associated to the rootfs
+func (stateMachine *StateMachine) findRootfsVolumeStructure() (*gadget.Volume, string, *gadget.VolumeStructure) {
 	for volumeName, volume := range stateMachine.GadgetInfo.Volumes {
-		for _, structure := range volume.Structure {
-			if structure.Size == 0 {
-				return volume, volumeName
+		for i := range volume.Structure {
```

Comment:
Any reason we don't do `for _, s := range volume.Structure {` here as it was done previously? Since we don't really care about the index, only the structure.